rclpy/rclpy/lifecycle.py

```diff
@@ -0,0 +1,408 @@
+# Copyright 2021 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import wraps
+from typing import Callable
+from typing import Dict
+from typing import Optional
+from typing import Set
+
+import lifecycle_msgs.srv
+
+from rclpy.callback_groups import CallbackGroup
+from rclpy.impl.implementation_singleton import rclpy_implementation as _rclpy
+from rclpy.node import Node
+from rclpy.publisher import Publisher
+from rclpy.qos import QoSProfile
+from rclpy.service import Service
+from rclpy.type_support import check_is_valid_srv_type
+
+
+TransitionCallbackReturn = _rclpy.TransitionCallbackReturnType
+
+
+class ManagedEntity:
+
+    def on_configure(self, state) -> TransitionCallbackReturn:
+        """Handle configure transition request."""
+        return TransitionCallbackReturn.SUCCESS
+
+    def on_cleanup(self, state) -> TransitionCallbackReturn:
+        """Handle cleanup transition request."""
+        return TransitionCallbackReturn.SUCCESS
+
+    def on_shutdown(self, state) -> TransitionCallbackReturn:
+        """Handle shutdown transition request."""
+        return TransitionCallbackReturn.SUCCESS
+
+    def on_activate(self, state) -> TransitionCallbackReturn:
+        """Handle activate transition request."""
+        return TransitionCallbackReturn.SUCCESS
+
+    def on_deactivate(self, state) -> TransitionCallbackReturn:
+        """Handle deactivate transition request."""
+        return TransitionCallbackReturn.SUCCESS
+
+    def on_error(self, state) -> TransitionCallbackReturn:
+        """Handle error transition request."""
+        return TransitionCallbackReturn.SUCCESS
+
+
+class SimpleManagedEntity(ManagedEntity):
```

Comment:
Is there motivation to have the `SimpleManagedEntity` inherit from a `ManagedEntity` other than that they share two intended transitions (activate/deactivate). Presumably the other transitions do not do anything.

Reply:
I though in some cases it could be useful to override some of the others, so I left all of them.

e.g. you can inherit from LifecyclePublisher and override `on_error()`.
I'm not sure if that's really useful or not.

If the only use-case we see is something like `SimpleManagedEntity`, i.e. setting/clearing a flag in on_active/on_deactive, maybe it would be simpler to just keep a reference of the lifecycle node state machine in all managed entities and query if the current state is active or not.
That might be a bit slower, as there a few extra indirections and comparisons.